scripts/deploy.py

```diff
@@ -0,0 +1,497 @@
+            return False
+        return True
+
+    if doc:
+        f.__doc__ = doc
+    else:
+        f.__doc__ = "Execution of %s " % cmd_line
+    return f
+
+
+def RetryPrompt():
+    while True:
+        click.secho('(type one of: retry/ignore/abort)', fg='cyan')
+        value = click.prompt('', prompt_suffix='>>>>>>> ')
+        if value == 'retry':
+            return True
+        if value == 'ignore':
+            return False
+        if value == 'abort':
+            sys.exit(1)
+
+
+# Exception when step decides to jump to another step.
+class Jump(BaseException):
+    def __init__(self, whereto):
+        self.whereto = whereto
+
+
+class Pipeline:
+    def __init__(self):
+        self.steps = []  # List of steps in a pipeline.
+        self.start = 1  # From which step to start, 1-based.
+        self.end = None  # At which step to end.
+        self.list_only = False  # Only list steps, do not execute.
+        self.step_by_step = False  # Ask confirmation before every step.
+        self.context = {}
+        self.cmd_name = 'unknown'
+
+    def AddStep(self, func):
+        self.steps.append(func)
+
+    def StateFileName(self):
+        return os.path.join('/tmp', '%s_state.json' % self.cmd_name)
+
+    def MaybeLoadState(self):
+        if os.path.isfile(self.StateFileName()):
+            if click.confirm('Forgotten state found. Restore?'):
+                with open(self.StateFileName()) as f:
+                    self.context = json.loads(f.read())
+                    if 'chdir' in self.context:
+                        os.chdir(self.context['chdir'])
+
+    def StoreState(self):
+        with open(self.StateFileName(), 'w') as f:
+            f.write(json.dumps(self.context))
+
+    def Run(self, cmd_name):
+        self.cmd_name = cmd_name
+        if self.list_only:
+            for i, n in enumerate(self.steps):
+                click.secho('%2d. %s' % (i + 1, n.__doc__))
+            return
+        click.clear()
+
+        if self.end is None:
+            end = len(self.steps)
+        self.context['idx'] = self.start - 1
+
+        self.MaybeLoadState()
+
+        while self.start <= (self.context['idx'] + 1) <= end:
+            self.StoreState()
+
+            task_f = self.steps[self.context['idx']]
+            try:
+                click.echo(
+                    click.style(
+                        '[%2d/%2d]' %
+                        (self.context['idx'] + 1, len(self.steps)),
+                        fg='green') + ' %s...' % task_f.__doc__)
+                if self.step_by_step:
+                    if not click.confirm('Should I run it?'):
+                        raise click.Abort
+                if task_f(self.context):
+                    self.context['idx'] += 1
+                    continue
+            except click.Abort:
+                raise
+            except Jump as jmp:
+                self.context['idx'] += jmp.whereto
```

Comment:
`whereto` should be the absolute target, not a relative offset. If it is a relative offset (as the `+=` indicates), then it should be renamed to `offset` or similar

Reply:
Done.